gateway/reverse_proxy.go

```diff
@@ -803,38 +798,27 @@ type memConnProvider struct {
 }
 
 var memConnProviders = &struct {
-	mtx sync.Mutex
-	m   map[string]*memConnProvider
+	mtx           sync.Mutex
+	m             map[string]*memConnProvider
+	lastIdleCheck int64
 }{
 	m: make(map[string]*memConnProvider),
 }
 
-// cleanIdleMemConnProvidersEagerly deletes unused memconn.Provider instances and closes
+// cleanIdleMemConnProviders deletes unused memconn.Provider instances and closes
 // its listener.
-func cleanIdleMemConnProvidersEagerly(maxIdleDuration time.Duration) {
-	memConnProviders.mtx.Lock()
-	defer memConnProviders.mtx.Unlock()
-
+func cleanIdleMemConnProviders(maxIdleDuration, idleCheckInterval time.Duration) {
+	if memConnProviders.lastIdleCheck > time.Now().Add(-idleCheckInterval).Unix() {
```

Comment:
I'm imagining the outside loop to include the idleCheckInterval e.g.

```go
go func() {
    for {
       time.Sleep(checkIdleMemConnInterval)
       cleanIdleMemConnProviders(maxIdleMemConnDuration)
    }
}()```

Reply:
I'd honestly recommend a sync.Pool for this. It seems we can just drop half of this code, and let Pool take care of allocating (and freeing) the memConnProviders as the load rises/falls.

Reply:
> I'd honestly recommend a sync.Pool for this. It seems we can just drop half of this code, and let Pool take care of allocating (and freeing) the *memConnProvider-s as the load rises/falls.

We discussed this topic extensively on Slack. Here is my opinion:

> I think `sync.Pool` would be cool but the current implementation requires a running HTTP server in the provider. I couldn’t find any way to clean the goroutines and stop listeners.